QueueImplementation/VM/CustomCollectionViewModel.cs

```diff
@@ -0,0 +1,363 @@
+			}
+			set
+			{
+				if (this.isChoisePossible == value)
+					return;
+
+				this.isChoisePossible = value;
+				this.RaisePropertyChanged(() => this.IsChoiseAvailable);
+			}
+		}
+
+		/// <summary>
+		/// Gets or sets the message.
+		/// </summary>
+		/// <owner>Anton Petrenko</owner>
+		/// <value>The message.</value>
+		public string Message
+		{
+			get
+			{
+				return this.message;
+			}
+			set
+			{
+				this.message = value;
+				this.RaisePropertyChanged(() => this.Message);
+			}
+		}
+
+		/// <summary>
+		/// Gets the remove command.
+		/// </summary>
+		/// <owner>Anton Petrenko</owner>
+		/// <value>The remove command.</value>
+		public RelayCommand Remove
+		{
+			get
+			{
+				if (this.remove is null)
+					this.remove = new RelayCommand(this.RemoveNumber, this.CanGetNumber);
+
+				return this.remove;
+			}
+			set
+			{
+				if (this.remove == value)
+					return;
+
+				this.remove = value;
+				this.RaisePropertyChanged(() => this.Remove);
+			}
+		}
+
+		/// <summary>
+		/// Removes the number from the collection.
+		/// </summary>
+		/// <owner>Anton Petrenko</owner>
+		/// <param name="obj">The object.</param>
+		private void RemoveNumber(object obj)
+		{
+			try
+			{
+				this.RestoreCurrentElement();
+				var removedItem = this.Collection.Remove();
+				this.Message = $"The item {removedItem} was removed.";
+			}
+			catch (InvalidOperationException ex)
+			{
+				this.Message = ex.Message;
+			}
+
+			if (this.Collection.Count() == 0)
+			{
+				this.Message += " The collection is empty.";
+				this.RestoreRemoveAndShowCommands();
+			}
+
+			this.RaisePropertyChanged(() => this.AllNumbers);
+		}
+
+		/// <summary>
+		/// Restores the current element.
+		/// </summary>
+		/// <owner>Anton Petrenko</owner>
+		private void RestoreCurrentElement()
+		{
+			if (this.CurrentElement.HasValue)
+				this.CurrentElement = null;
+		}
+
+		/// <summary>
+		/// Restores the remove and show commands.
+		/// </summary>
+		/// <owner>Anton Petrenko</owner>
+		private void RestoreRemoveAndShowCommands()
+		{
+			this.Show = null;
+			this.Remove = null;
+		}
+
+		/// <summary>
+		/// Gets or sets the selected type of the collection (array/linked list).
+		/// </summary>
+		/// <owner>Anton Petrenko</owner>
+		/// <value>The type of the selected array linked list.</value>
+		public string SelectedArrayLinkedListType { get; set; }
+
+		/// <summary>
+		/// Gets or sets the selected type of the collection (queue/stack).
+		/// </summary>
+		/// <owner>Anton Petrenko</owner>
+		/// <value>The type of the selected queue stack.</value>
+		public string SelectedQueueStackType { get; set; }
+
+		/// <summary>
+		/// Gets and sets the show command.
+		/// </summary>
+		/// <owner>Anton Petrenko</owner>
+		/// <value>The show command.</value>
+		public RelayCommand Show
+		{
+			get
+			{
+				if (this.show is null)
+					this.show = new RelayCommand(this.ShowNumber, this.CanGetNumber);
+
+				return this.show;
+			}
+			set
```

Comment:
I think setter is excessive for the commands. But to make controls inactive we should handle that using CanGet* delegate.

Reply:
Fixed.

Reply:
But you also removed the caching of the commands. And this should be reverted.

Reply:
Restored